src/detect-tls-cert.c

```diff
@@ -0,0 +1,496 @@
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Mats Klepsland <mats.klepsland@gmail.com>
+ *
+ * Implements support for tls_cert keyword.
+ */
+
+#include "suricata-common.h"
+#include "threads.h"
+#include "debug.h"
+#include "decode.h"
+#include "detect.h"
+
+#include "detect-parse.h"
+#include "detect-engine.h"
+#include "detect-engine-mpm.h"
+#include "detect-engine-prefilter.h"
+#include "detect-content.h"
+#include "detect-pcre.h"
+#include "detect-tls-cert.h"
+
+#include "flow.h"
+#include "flow-util.h"
+#include "flow-var.h"
+
+#include "util-debug.h"
+#include "util-unittest.h"
+#include "util-spm.h"
+#include "util-print.h"
+
+#include "stream-tcp.h"
+
+#include "app-layer.h"
+#include "app-layer-ssl.h"
+
+#include "util-unittest.h"
+#include "util-unittest-helper.h"
+
+static int DetectTlsCertSetup(DetectEngineCtx *, Signature *, const char *);
+static void DetectTlsCertRegisterTests(void);
+static InspectionBuffer *GetData(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms,
+        Flow *_f, const uint8_t _flow_flags,
+        void *txv, const int list_id);
+static int g_tls_cert_buffer_id = 0;
+
+/**
+ * \brief Registration function for keyword: tls_cert
+ */
+void DetectTlsCertRegister(void)
+{
+    sigmatch_table[DETECT_AL_TLS_CERT].name = "tls_cert";
+    sigmatch_table[DETECT_AL_TLS_CERT].desc = "content modifier to match the TLS certificate buffer";
+    sigmatch_table[DETECT_AL_TLS_CERT].url = DOC_URL DOC_VERSION "/rules/tls-keywords.html#tls-cert";
+    sigmatch_table[DETECT_AL_TLS_CERT].Match = NULL;
+    sigmatch_table[DETECT_AL_TLS_CERT].Setup = DetectTlsCertSetup;
+    sigmatch_table[DETECT_AL_TLS_CERT].Free = NULL;
+    sigmatch_table[DETECT_AL_TLS_CERT].RegisterTests = DetectTlsCertRegisterTests;
+
+    sigmatch_table[DETECT_AL_TLS_CERT].flags |= SIGMATCH_NOOPT;
+
+    DetectAppLayerInspectEngineRegister2("tls_cert", ALPROTO_TLS,
+            SIG_FLAG_TOCLIENT, TLS_STATE_CERT_READY,
+            DetectEngineInspectBufferGeneric, GetData);
+
+    DetectAppLayerMpmRegister2("tls_cert", SIG_FLAG_TOCLIENT, 2,
+            PrefilterGenericMpmRegister, GetData, ALPROTO_TLS,
+            TLS_STATE_CERT_READY);
+
+    DetectBufferTypeSetDescriptionByName("tls_cert", "TLS certificate");
+
+    g_tls_cert_buffer_id = DetectBufferTypeGetByName("tls_cert");
+}
+
+/**
+ * \brief This function setup the tls_cert modifier keyword
+ *
+ * \param de_ctx Pointer to the Detect Engine Context
+ * \param s      Pointer to the Signature to which the keyword belongs
+ * \param str    Should hold an empty string always
+ *
+ * \retval 0 On success
+ * \retval -1 on failure
+ */
+static int DetectTlsCertSetup(DetectEngineCtx *de_ctx, Signature *s,
+                              const char *str)
+{
+    DetectBufferSetActiveList(s, g_tls_cert_buffer_id);
```

Comment:
check the return code here as well

Reply:
Will do!

Reply:
This is probably something that is missing from a lot of other places too. So maybe you can check at least all the TLS ones for this?